src/lib/ReactCore_Internal.re

```diff
@@ -403,64 +462,53 @@ module Make = (Implementation: HostImplementation) => {
         };
       let stateNotUpdated = opaqueInstance === updatedOpaqueInstance;
       let bailOut = {
-        let NativeInstance(_, {element}) | Instance({element}) = opaqueInstance;
-        stateNotUpdated && element === nextElement
+        let Instance({element}) = opaqueInstance;
+        stateNotUpdated && element === nextElement;
       };
       let logUpdate =
-          (
-            ~componentChanged,
-            ~stateChanged,
-            ~subTreeChanged,
-            newOpaqueInstance
-          ) => {
-        let NativeInstance(_, {id}) | Instance({id}) = opaqueInstance;
-        let NativeInstance(_, {id: newId}) | Instance({id: newId}) = newOpaqueInstance;
-        let comp = componentChanged ? " component" : "";
-        let st = stateChanged ? " state" : "";
-        let sub = subTreeChanged ? " subtree" : "";
+          (~stateChanged, ~subTreeChanged, oldInstance, newInstance) => {
+        let {id} = oldInstance;
+        let {id: newId} = newInstance;
         UpdateLog.add(
           updateLog,
           UpdateLog.UpdateInstance({
             oldId: id,
             newId,
-            oldOpaqueInstance: opaqueInstance,
-            newOpaqueInstance,
-            componentChanged,
+            instanceUpdate: InstanceUpdate(oldInstance, newInstance),
             stateChanged,
             subTreeChanged
           })
         );
-        logString(
-          Printf.sprintf(
-            "instance #%d updated to instance #%d changed:%s%s%s",
-            id,
+      };
+      let logComponentSwitch = (~subTreeChanged, newOpaqueInstance) => {
+        let Instance({id}) = opaqueInstance;
+        let Instance({id: newId}) = newOpaqueInstance;
+        UpdateLog.add(
+          updateLog,
+          UpdateLog.SwitchComponent({
+            oldId: id,
             newId,
-            comp,
-            st,
-            sub
-          )
-        )
+            oldOpaqueInstance: opaqueInstance,
+            newOpaqueInstance,
+            subTreeChanged
+          })
+        );
       };
       if (bailOut) {
-        opaqueInstance
+        opaqueInstance;
       } else {
-        let resetHandedOffInstance =
-          switch updatedOpaqueInstance {
-          | NativeInstance(_, {component} as updatedInstance) =>
-            component.handedOffInstance := Some(updatedInstance);
-            (() => component.handedOffInstance := None)
-          | Instance({component} as updatedInstance) =>
-            component.handedOffInstance := Some(updatedInstance);
-            (() => component.handedOffInstance := None)
-          };
+        let resetHandedOffInstance = {
```

Comment:
I'd remove this function and just do  `component.handedOffInstance := None` inside the switch.